src/exchange/withdrawal_handler.cairo

```diff
@@ -182,7 +308,60 @@ mod WithdrawalHandler {
         /// * `reason_bytes` - The reason of the error.
         fn handle_withdrawal_error(
             ref self: ContractState, key: felt252, starting_gas: u128, reason_bytes: Array<felt252>
-        ) { // TODO
+        ) {
+            // Just cancels withdrawal. There is no way to handle revert and revert reason right now.
+
+            withdrawal_utils::cancel_withdrawal(
+                self.data_store.read(),
+                self.event_emitter.read(),
+                self.withdrawal_vault.read(),
+                key,
+                get_caller_address(),
+                starting_gas,
+                '', // TODO: There is no way to get revert message currently.
+                reason_bytes
+            );
+        }
+
+        /// Executes a withdrawal with keeper.
+        /// # Arguments
+        /// * `key` - The key of the withdrawal to execute.
+        /// * `oracle_params` - The oracle params to set prices before execution.
+        /// * `keeper` - The keeper executing the withdrawal.
+        fn execute_withdrawal_keeper(
```

Comment:
This is marked as external in GMX code

Reply:
I don't think there is a way to call contract itself with a msg.sender context of this contract. In GMX there is a check msg.sender == address(this) so as I know it's not possible to achieve this currently in Starknet. so I moved this function to internal

Reply:
Yeah you're right didn't see that !